File: 2024/Day_06/part01.py
import numpy as np

# makes numpy matrix of chars
with open('test.txt', 'r') as file:
    data = np.array([list(line.strip()) for line in file])

# find location of ^
def init_loc(data):
    i, j = np.where(data == '^')
    return i[0], j[0]


# find marching path and returns num distinct positions
def find_path(i, j, data):
    distinct_pos = 0
    dir = [1, 2, 3, 4]  # [up, right, down, left]
    dir_pos = 0

    while(True):
        # going up
        if (dir_pos % len(dir)) == 0:
            for idx in range(i, -1, -1):
                i = idx
                if data[i][j] == '.':
                    data[i][j] = 'X'
                    distinct_pos += 1
                elif data[i][j] == '#':
                    i += 1              # go back 1
                    break
            if i == 0: return distinct_pos
        # going right
        elif (dir_pos % len(dir)) == 1:
            for idx in range(j, len(data)):
                j = idx
                if data[i][j] == '.':
                    data[i][j] = 'X'
                    distinct_pos += 1
                elif data[i][j] == '#':
                    j -= 1              # go back 1
                    break
            if j == len(data)-1: return distinct_pos
        # going down
        elif (dir_pos % len(dir)) == 2:
            for idx in range(i, len(data)):
                i = idx
                if data[i][j] == '.':
                    data[i][j] = 'X'
                    distinct_pos += 1
                elif data[i][j] == '#':
                    i -= 1              # go back 1
                    break
            if i == len(data)-1: return distinct_pos
        # going left
        else:
            for idx in range(j, -1, -1):
                j = idx
                if data[i][j] == '.':
                    data[i][j] = 'X'
                    distinct_pos += 1
                elif data[i][j] == '#':
                    j += 1              # go back 1
                    break
            if j == 0: return distinct_pos
        dir_pos += 1

i, j = init_loc(data)
distinct_pos = find_path(i, j, data)
print(distinct_pos+1)

File: 2024/Day_06/part01.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

void read_file(vector<vector<char>> &data) {
    ifstream file("test.txt");
    if (!file.is_open()) {
        cerr << "Error Opening File" << endl;
        exit(-1);       
    }

    string line;
    while (getline(file, line)) {
        vector<char> row(line.begin(), line.end());
        data.push_back(row);
    }
    file.close();
}

// find location of '^'
pair<int, int> init_loc(const vector<vector<char>>& data) {
    for (int i = 0; i < data.size(); ++i) {
        for (int j = 0; j < data[i].size(); ++j) {
            if (data[i][j] == '^') {
                return {i, j};
            }
        }
    }
    return {-1, -1};
}

// find marching path and returns num distinct positions
int find_path(int i, int j, vector<vector<char>>& data) {
    int distinct_pos = 0;
    int dir_pos = 0;
    vector<int> dir = {1, 2, 3, 4};  // [up, right, down, left]

    while (true) {
        // Going up
        if (dir_pos % dir.size() == 0) {
            for (int idx = i; idx >= 0; idx--) {
                i = idx;
                if (data[i][j] == '.') {
                    data[i][j] = 'X';
                    distinct_pos++;
                } else if (data[i][j] == '#') {
                    i++;  // Go back 1
                    break;
                }
            }
            if (i == 0) return distinct_pos;
        }
        // Going right
        else if (dir_pos % dir.size() == 1) {
            for (int idx = j; idx < data[i].size(); idx++) {
                j = idx;
                if (data[i][j] == '.') {
                    data[i][j] = 'X';
                    distinct_pos++;
                } else if (data[i][j] == '#') {
                    j--;  // Go back 1
                    break;
                }
            }
            if (j == data[i].size() - 1) return distinct_pos;
        }
        // Going down
        else if (dir_pos % dir.size() == 2) {
            for (int idx = i; idx < data.size(); idx++) {
                i = idx;
                if (data[i][j] == '.') {
                    data[i][j] = 'X';
                    distinct_pos++;
                } else if (data[i][j] == '#') {
                    i--;  // Go back 1
                    break;
                }
            }
            if (i == data.size() - 1) return distinct_pos;
        }
        // Going left
        else {
            for (int idx = j; idx >= 0; idx--) {
                j = idx;
                if (data[i][j] == '.') {
                    data[i][j] = 'X';
                    distinct_pos++;
                } else if (data[i][j] == '#') {
                    j++;  // Go back 1
                    break;
                }
            }
            if (j == 0) return distinct_pos;
        }
        dir_pos++;
    }
}

int main() {
    vector<vector<char>> data;
    read_file(data);
    auto [i, j] = init_loc(data);
    int distinct_pos = find_path(i, j, data);
    cout << distinct_pos + 1 << endl;

    return 0;
}
